From 0da843e7174a1d71825e50f381ad1829163626c7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Sat, 14 Dec 2024 17:31:58 +0100
Subject: [PATCH 80/86] wined3d: Avoid threads for now, still fails.

dlls/d2d1/tests/x86_64-windows/d2d1.ok
dlls/d3d10core/tests/x86_64-windows/d3d10core.ok
dlls/d3d11/tests/x86_64-windows/d3d11.ok
dlls/dxgi/tests/x86_64-windows/dxgi.ok
dlls/evr/tests/x86_64-windows/evr.ok
dlls/mfplat/tests/x86_64-windows/mfplat.ok

/home/bernhard/data/entwicklung/2024/wine/wine/tools/runtest -q -P wine -T . -M d2d1.dll -p dlls/d2d1/tests/x86_64-windows/d2d1_test.exe d2d1 && touch dlls/d2d1/tests/x86_64-windows/d2d1.ok
05dc:fixme:d3d:wined3d_guess_card No card selector available for card vendor 0000 (using GL_RENDERER "llvmpipe (LLVM 15.0.6, 256 bits)").
05dc:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_PERFORMANCE_INFORMATION
05dc:fixme:dxgi:DXGID3D10CreateDevice Ignoring flags 0x20.
05dc:fixme:dwrite:dwritefactory_CreateMonitorRenderingParams (0000000000000001): monitor setting ignored
05dc:fixme:d3d:state_linepattern_w Setting line patterns is not supported in OpenGL core contexts.
05dc:fixme:d3d11:d3d11_device_context_SwapDeviceContextState D3D10 interface emulation not fully implemented yet!
=================================================================
==1492==ERROR: AddressSanitizer: heap-use-after-free on address 0x7f2a74121798 at pc 0x6ffff4c053b6 bp 0x7ffffe1ff300 sp 0x7ffffe1ff348
READ of size 8 at 0x7f2a74121798 thread T0
0308:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff4c053b5 in wined3d_context_gl_update_window /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/context_gl.c:1338:56
    #1 0x6ffff4c053b5 in wined3d_context_gl_activate /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/context_gl.c:4329:5
    #2 0x6ffff4c04d94 in wined3d_context_gl_acquire /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/context_gl.c:4399:5
    #3 0x6ffff4d78777 in context_acquire /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/wined3d_private.h:4751:12
    #4 0x6ffff4d78777 in texture_resource_unload /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/texture.c:3389:15
    #5 0x6ffff4c41cd0 in wined3d_cs_st_submit /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/cs.c:2988:9
    #6 0x6ffff4c520cb in wined3d_device_uninit_3d /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/device.c:1739:9
    #7 0x6ffff4d52bba in wined3d_swapchain_decref /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/swapchain.c:158:13
    #8 0x6ffff83b6f90 in dxgi_device_Release /home/bernhard/data/entwicklung/2024/wine\wine/dlls/dxgi/device.c:98:9
    #9 0x0001400e22ab in IDXGIDevice_Release /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\dxgi.h:2110:12
    #10 0x0001400e22ab in release_test_context_ /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:1281:11
    #11 0x00014000a720 in test_clip /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:1890:5
    #12 0x00014000633b in run_queued_tests /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:456:13
    #13 0x00014000633b in func_d2d1 /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:15540:5
    #14 0x0001400fbb03 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #15 0x0001400fb54b in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h
    #16 0x0001400fd4df in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #17 0x6ffffc1c4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #18 0x6ffffc39fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

0x7f2a74121798 is located 1816 bytes inside of 1824-byte region [0x7f2a74121080,0x7f2a741217a0)
freed by thread T0 here:
    #0 0x6ffffa57a041 in free /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:71:3
    #1 0x6ffff4d52ca7 in wined3d_swapchain_decref /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/swapchain.c:162:9
    #2 0x6ffff83cbbee in d3d11_swapchain_Release /home/bernhard/data/entwicklung/2024/wine\wine/dlls/dxgi/swapchain.c:243:9
    #3 0x0001400e2245 in IDXGISwapChain_Release /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\dxgi.h:1737:12
    #4 0x0001400e2245 in release_test_context_ /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:1279:5
    #5 0x00014000a720 in test_clip /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:1890:5
    #6 0x00014000633b in run_queued_tests /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:456:13
    #7 0x00014000633b in func_d2d1 /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:15540:5
    #8 0x0001400fbb03 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #9 0x0001400fb54b in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h
    #10 0x0001400fd4df in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #11 0x6ffffc1c4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #12 0x6ffffc39fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

previously allocated by thread T0 here:
    #0 0x6ffffa57a276 in calloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:91:3
    #1 0x6ffff4bb797b in adapter_gl_create_swapchain /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/adapter_gl.c:4425:26
    #2 0x6ffff4d57006 in wined3d_swapchain_create /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/swapchain.c:1740:9
    #3 0x6ffff83c83e5 in d3d11_swapchain_init /home/bernhard/data/entwicklung/2024/wine\wine/dlls/dxgi/swapchain.c:1020:9
    #4 0x6ffff83b986c in dxgi_swapchain_factory_create_swapchain /home/bernhard/data/entwicklung/2024/wine\wine/dlls/dxgi/device.c:464:9
    #5 0x6ffff83bdf38 in IWineDXGISwapChainFactory_create_swapchain /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include/wine\winedxgi.h:146:12
    #6 0x6ffff83bdf38 in dxgi_factory_CreateSwapChainForHwnd /home/bernhard/data/entwicklung/2024/wine\wine/dlls/dxgi/factory.c:302:14
    #7 0x6ffff83bd6cd in IWineDXGIFactory_CreateSwapChainForHwnd /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include/wine\winedxgi.h:1024:12
    #8 0x6ffff83bd6cd in dxgi_factory_CreateSwapChain /home/bernhard/data/entwicklung/2024/wine\wine/dlls/dxgi/factory.c:238:12
    #9 0x0001400e2e42 in IDXGIFactory_CreateSwapChain /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\dxgi.h:1958:12
    #10 0x0001400e2e42 in create_swapchain /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:1211:10
    #11 0x0001400e0c02 in init_test_context_ /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:1302:22
    #12 0x000140007de7 in test_clip /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:1721:10
    #13 0x00014000633b in run_queued_tests /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:456:13
    #14 0x00014000633b in func_d2d1 /home/bernhard/data/entwicklung/2024/wine\wine/dlls/d2d1/tests/d2d1.c:15540:5
    #15 0x0001400fbb03 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #16 0x0001400fb54b in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h
    #17 0x0001400fd4df in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #18 0x6ffffc1c4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #19 0x6ffffc39fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

SUMMARY: AddressSanitizer: heap-use-after-free /home/bernhard/data/entwicklung/2024/wine\wine/dlls/wined3d/context_gl.c:1338:56 in wined3d_context_gl_update_window
Shadow bytes around the buggy address:
  0x7f2a74121500: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f2a74121580: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f2a74121600: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f2a74121680: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f2a74121700: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x7f2a74121780: fd fd fd[fd]fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f2a74121800: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x7f2a74121880: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f2a74121900: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f2a74121980: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x7f2a74121a00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1492==ABORTING
make: *** [Makefile:41397: dlls/d2d1/tests/x86_64-windows/d2d1.ok] Fehler 1
0308:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
---
 dlls/d2d1/tests/d2d1.c           | 2 +-
 dlls/d3d10core/tests/d3d10core.c | 2 +-
 dlls/d3d11/tests/d3d11.c         | 2 +-
 dlls/dxgi/swapchain.c            | 2 +-
 dlls/dxgi/tests/dxgi.c           | 2 +-
 dlls/wined3d/wined3d_main.c      | 2 +-
 6 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/dlls/d2d1/tests/d2d1.c b/dlls/d2d1/tests/d2d1.c
index 625ad53292f..f9e16c4cb64 100644
--- a/dlls/d2d1/tests/d2d1.c
+++ b/dlls/d2d1/tests/d2d1.c
@@ -15451,7 +15451,7 @@ START_TEST(d2d1)
     pD2D1Vec3Length = (void *)GetProcAddress(d2d1_dll, "D2D1Vec3Length");
     pD2D1ConvertColorSpace = (void *)GetProcAddress(d2d1_dll, "D2D1ConvertColorSpace");
 
-    use_mt = !getenv("WINETEST_NO_MT_D3D");
+    use_mt = 0;//!getenv("WINETEST_NO_MT_D3D");
     /* Some host drivers (MacOS, Mesa radeonsi) never unmap memory even when
      * requested. When using the chunk allocator, running the tests with more
      * than one thread can exceed the 32-bit virtual address space. */
diff --git a/dlls/d3d10core/tests/d3d10core.c b/dlls/d3d10core/tests/d3d10core.c
index 9cb0ec118a4..dec69121ec8 100644
--- a/dlls/d3d10core/tests/d3d10core.c
+++ b/dlls/d3d10core/tests/d3d10core.c
@@ -19722,7 +19722,7 @@ START_TEST(d3d10core)
             damavand = true;
     }
 
-    use_mt = !getenv("WINETEST_NO_MT_D3D");
+    use_mt = 0;//!getenv("WINETEST_NO_MT_D3D");
     /* Some host drivers (MacOS, Mesa radeonsi) never unmap memory even when
      * requested. When using the chunk allocator, running the tests with more
      * than one thread can exceed the 32-bit virtual address space. */
diff --git a/dlls/d3d11/tests/d3d11.c b/dlls/d3d11/tests/d3d11.c
index 9c5b35faafa..7a51c8016d6 100644
--- a/dlls/d3d11/tests/d3d11.c
+++ b/dlls/d3d11/tests/d3d11.c
@@ -36597,7 +36597,7 @@ START_TEST(d3d11)
             damavand = true;
     }
 
-    use_mt = !getenv("WINETEST_NO_MT_D3D");
+    use_mt = 0;//!getenv("WINETEST_NO_MT_D3D");
     /* Some host drivers (MacOS, Mesa radeonsi) never unmap memory even when
      * requested. When using the chunk allocator, running the tests with more
      * than one thread can exceed the 32-bit virtual address space. */
diff --git a/dlls/dxgi/swapchain.c b/dlls/dxgi/swapchain.c
index 4797fcecd8d..3aa9d752f73 100644
--- a/dlls/dxgi/swapchain.c
+++ b/dlls/dxgi/swapchain.c
@@ -1008,7 +1008,7 @@ HRESULT d3d11_swapchain_init(struct d3d11_swapchain *swapchain, struct dxgi_devi
 
     swapchain->IDXGISwapChain4_iface.lpVtbl = &d3d11_swapchain_vtbl;
     swapchain->state_parent.ops = &d3d11_swapchain_state_parent_ops;
-    swapchain->refcount = 1;
+    swapchain->refcount = 1; /* setting to 2 would avoid the free, but tests would fail because of this */
     wined3d_mutex_lock();
     wined3d_private_store_init(&swapchain->private_store);
 
diff --git a/dlls/dxgi/tests/dxgi.c b/dlls/dxgi/tests/dxgi.c
index 912bcd8fa61..db8cfa51cd0 100644
--- a/dlls/dxgi/tests/dxgi.c
+++ b/dlls/dxgi/tests/dxgi.c
@@ -8018,7 +8018,7 @@ START_TEST(dxgi)
     registry_mode.dmSize = sizeof(registry_mode);
     ok(EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &registry_mode), "Failed to get display mode.\n");
 
-    use_mt = !getenv("WINETEST_NO_MT_D3D");
+    use_mt = 0;//!getenv("WINETEST_NO_MT_D3D");
     /* Some host drivers (MacOS, Mesa radeonsi) never unmap memory even when
      * requested. When using the chunk allocator, running the tests with more
      * than one thread can exceed the 32-bit virtual address space. */
diff --git a/dlls/wined3d/wined3d_main.c b/dlls/wined3d/wined3d_main.c
index 3b436919559..02599ac0bb4 100644
--- a/dlls/wined3d/wined3d_main.c
+++ b/dlls/wined3d/wined3d_main.c
@@ -115,7 +115,7 @@ CRITICAL_SECTION wined3d_command_cs = {&wined3d_command_cs_debug, -1, 0, 0, 0, 0
  * where appropriate. */
 struct wined3d_settings wined3d_settings =
 {
-    .cs_multithreaded = WINED3D_CSMT_ENABLE,
+    .cs_multithreaded = 0/*WINED3D_CSMT_ENABLE*/,
     .max_gl_version = MAKEDWORD_VERSION(4, 4),
     .pci_vendor_id = PCI_VENDOR_NONE,
     .pci_device_id = PCI_DEVICE_NONE,
-- 
2.47.1

